### PR TITLE
resume hyperopt run

### PR DESCRIPTION
**Pull request by shusso**
## Summary
Resume hyperopt based on previous hyperopt run 

## What's new?
Hyperopt will now handle SIGINT and upon receiving it, it'll pickle the current trials file. When hyperopt is ran the next time, it'll un-pickle the trials file and continue from previous run.

### CODE REVIEW

File: freqtrade/optimize/hyperopt.py [hunk 1]
```diff
@@ -43,6 +46,10 @@
 PROCESSED = None  # optimize.preprocess(optimize.load_data())
 OPTIMIZE_CONFIG = hyperopt_optimize_conf()
 
+# Hyperopt Trials
+TRIALS_FILE = 'freqtrade/optimize/hyperopt_trials.pickle'
```
Comment (gcarq) [inline freqtrade/optimize/hyperopt.py]: This won't work on windows. You can use path.join

File: freqtrade/optimize/hyperopt.py [hunk 2]
```diff
@@ -99,6 +106,26 @@
 }
 
 
+def save_trials(trials, trials_path=TRIALS_FILE):
+    "Save hyperopt trials to file"
```
Comment (gcarq) [inline freqtrade/optimize/hyperopt.py]: A docstring should be enquoted with triple quotes, e.g.:
```
def kos_root():
    """Return the pathname of the KOS root directory."""
    global _kos_root
    if _kos_root: return _kos_root
    ...
```
Same for some other functions in this PR

File: freqtrade/tests/optimize/test_hyperopt.py [hunk 3]
```diff
@@ -27,16 +26,37 @@ def test_loss_calculation_has_limited_profit():
...
 
 def create_trials(mocker):
+    """
+    When creating trials, mock the hyperopt Trials so that *by default*
+      - we don't create any pickle'd files in the filesystem
+      - we might have a pickle'd file so make sure that we return
+        false when looking for it
+    """
+    mocker.patch('freqtrade.optimize.hyperopt.TRIALS_FILE',
+                 return_value='freqtrade/tests/optimize/ut_trials.pickle')
```
Comment (gcarq) [inline freqtrade/tests/optimize/test_hyperopt.py]: Same as above, path.join would be awesome

**Comment by shusso** [comment]
> It will allow users to change the path without touching to core files.

I'm wondering is there a real need to do it? I'd rather merge this as it is, and if there is a need to change it we can do it imho

**Comment by ermakus** [comment]
I am slightly improved log_results in my fork to preserve space_eval formatting and reuse it in start():
```
def log_trials_result(trials):

    def arr_to_scalar(v):
        return v[0] if len(v) else 0

    try:
        vals = trials.best_trial['misc']['vals']
        # Convert to format accepted by space_eval
        vals = {k: arr_to_scalar(v) for k, v in vals.items()}
        best_parameters = space_eval(SPACE, vals)
        best_result = trials.best_trial['result']['result']
    except ValueError:
        best_parameters = {}
        best_result = 'Sorry, Hyperopt was not able to find good parameters. Please ' \
                      'try with more epochs (param: -e).'
    logger.info('Best parameters:\n%s', json.dumps(best_parameters, indent=4))
    logger.info('Best Result:\n%s', best_result)

def start(args):
...
    try:
        fmin(
            fn=optimizer,
            space=SPACE,
            algo=tpe.suggest,
            max_evals=TOTAL_TRIES,
            trials=TRIALS
        )
    except (ValueError, KeyboardInterrupt):
        pass

    log_trials_result(TRIALS)
```

**Comment by glonlas** [comment]
@shusso of course merge it, we can easily update it later if needed :)

**Comment by glonlas** [comment]
@ermakus nice :+1:

**Pull request merged**